R/estimate_time_varying.R

```diff
@@ -0,0 +1,187 @@
+#' smooth the case and death time-series, using a moving average procedure
+#' before calculating the time-varying severity. Useful for noisey time-series
+#' or time-series with strong reporting (e.g., weekend) effects
+#'
+#' @param smoothing_window A number determining the smoothing window size to use
+#' when smoothing the case and death time-series, using a moving average
+#' procedure before calculating the time-varying severity.
+#' Useful for noisy time-series or time-series with strong reporting
+#' (e.g., weekend) effects.
+#' The default value is 1 for _no smoothing_. Values > 1 apply smoothing.
+#'
+#' @param correct_for_delays A boolean flag indicating whether the user wishes
+#' to correct for the delay between case detection and death. FALSE corresponds
+#' to a naive severity being calculated, TRUE corresponds to the user
+#' calculating a corrected severity
+#'
+#' @return A data.frame containing the MLE estimate and 95% confidence interval
+#' of the corrected severity
+#'
+#' @export
+#'
+#' @examples
+#'
+#' library(datadelay)
+#' library(epiparameter)
+#' library(covidregionaldata)
+#' library(dplyr)
+#'
+#' df_covid_uk <- get_national_data(
+#'   countries = "united kingdom", source = "who", verbose = FALSE
+#' )
+#' df_covid_uk <- dplyr::rename(
+#'   df_covid_uk,
+#'   cases = cases_new, deaths = deaths_new
+#' )
+#'
+#' df_covid_uk_subset <- subset(df_covid_uk, date <= "2020-12-31")
+#'
+#' onset_to_death_covid <- epidist_db(
+#'   disease = "COVID-19",
+#'   epi_dist = "onset_to_death",
+#'   author = "Linton_etal"
+#' )
+#'
+#' df_covid_cfr_uk_naive <- estimate_time_varying(
+#'   df_covid_uk_subset,
+#'   epi_dist = onset_to_death_covid,
+#'   smooth_inputs = TRUE,
+#'   burn_in_value = 7L,
+#'   correct_for_delays = FALSE
+#' )
+#'
+#' df_covid_cfr_uk_corrected <- estimate_time_varying(
+#'   df_covid_uk_subset,
+#'   epi_dist = onset_to_death_covid,
+#'   smooth_inputs = TRUE,
+#'   burn_in_value = 7L,
+#'   correct_for_delays = TRUE
+#' )
+#'
+estimate_time_varying <- function(df_in,
+                                  epi_dist,
+                                  burn_in_value = get_default_burn_in(epi_dist),
+                                  smooth_inputs = FALSE,
+                                  smoothing_window = 7,
+                                  correct_for_delays = TRUE) {
+  # TODO input checking
+  checkmate::assert_logical(smooth_inputs, len = 1L, any.missing = FALSE)
+  checkmate::assert_logical(correct_for_delays, len = 1L, any.missing = FALSE)
+  checkmate::assert_integerish(burn_in_value, lower = 0, len = 1L)
+
+  pmf_vals <- stats::density(
+    epi_dist,
+    at = seq(from = 0, to = nrow(df_in) - 1L)
+  )
+
+  # smooth cases if requested
+  if (smooth_inputs) {
+    df_in$cases <- round(
+      zoo::rollmean(df_in$cases, k = smoothing_window, fill = NA)
+    )
+
+    df_in$deaths <- round(
+      zoo::rollmean(df_in$deaths, k = smoothing_window, fill = NA)
+    )
```

Comment:
Suggest replacing `zoo::rollmean()` with `stats::runmed()` - the median is less sensitive to outliers that might result from things like weekend effects, and this also avoids the dependency on {zoo}.